---
title: "CHOPOFF GUIDE"
author: "Kornel Labun & Haakon Tjeldnes"
date: "`r BiocStyle::doc_date()`"
package: "`r pkg_ver('crisprCHOPOFF')`"
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{CHOPOFF GUIDE}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Introduction

Welcome to crisprCHOPOFF, the R wrapper for CHOPOFF.

## Overview
This vignette shows 5 examples:

1. The Cas9 run returning GRangesList (most friendly)
2. A complete Step by step Run of Cas9 guides
3. A short Run of Cas12a guides
4. A short example in crisprVerse syntax, comparing bowtie to CHOPOFF 
 for finding a justified candidate guide in a selected gene.
5. A Full example in crisprVerse syntax, comparing bowtie to CHOPOFF 
 for finding a justified candidate guidein a selected gene, 
 with some additional analysis.

## Simple run CAS9

Here we will run CAS9 preset on sample guides, showing output as GRangesList
```{r simple_cas9}
library(crisprCHOPOFF)

name <- "CAS9"
genome <- system.file("extdata/sample_genome", "semirandom.fa", package = "crisprCHOPOFF")

out_dir_index <- file.path(tempdir(), "CHOPOFF_sample_genome")
build_index(name, genome, out_dir_index, distance = 2)
#'
# Now search some guides
guides <- system.file("extdata/sample_genome", "guides.txt", package = "crisprCHOPOFF")
guideSet <- addSpacerAlignmentsCHOPOFF(guides, out_dir_index, distance = 2, crisprDesign_message = FALSE)
guideSet$alignments[1:2] # guide (protospacer) 2, has 2 alignments (spacers) (on chromosome 1 & 5)
```
 
## Step by step run CAS9

Here we will run CAS9 preset on sample guides, showing it step by step.
```{r full_cas9}
library(crisprCHOPOFF)

name <- "CAS9"
genome <- system.file("extdata/sample_genome", "semirandom.fa", package = "crisprCHOPOFF")
## Note: a fasta index ".fai" file must exist in directory of genome.
# You can make it with:
#if (!file.exists(paste0(genome, ".fai"))) {
#Rsamtools::indexFa(genome)
#}
out_dir_index <- file.path(tempdir(), "CHOPOFF_sample_genome")
build_index(name, genome, out_dir_index, distance = 2)
#'
# Now search some guides
guides <- system.file("extdata/sample_genome", "guides.txt", package = "crisprCHOPOFF")
# Quick preview in guides:
guide_candidates <- read.table(guides, col.names = "guides")
unique(nchar(unlist(guide_candidates))) # Unique lengths of guides
guide_hits <- search_index(guides, out_dir_index, distance = 2, validate = FALSE)
guide_hits_table <- read.table(guide_hits, sep = ",", header = TRUE)
# use data.table::fread for reading in large list
# Subset to 0 distance hits
dist0 <- guide_hits_table[guide_hits_table$distance == 0,]
head(dist0)
# Which chromosomes is a specific guide found on with 0 distance hits?
unique(dist0[dist0$guide == "TCCGGCCTGGTTATCGAAGG",]$chromosome) # 2 chromosomes

# Summarize overlaps
summarize_path <- summarize_overlaps(guide_hits, distance = 2, validate = FALSE)
summarize_table <- read.table(summarize_path, sep = ",", header = TRUE)
sum(summarize_table$D2 > 0) # 10 guides have distance 2 alignments

# Filter overlaps
filter_path <- filter_overlaps(guide_hits, distance = 2, validate = FALSE)
filter_table <- read.table(filter_path, sep = ",", header = TRUE)
```

## Short Cas12a example

```{r simple_cas12a}
name <- "CAS12a"
out_dir_index <- file.path(tempdir(), "CHOPOFF_sample_genome", name)
build_index(name, genome, out_dir_index, distance = 2, preset = "Cas12a", validate = FALSE)
#'
# Now search some guides
guides_cas12 <- system.file("extdata/sample_genome", "guides_cas12a.txt", package = "crisprCHOPOFF")
# Quick preview in guides:
guide_candidates <- read.table(guides_cas12, col.names = "guides")
unique(nchar(unlist(guide_candidates))) # Unique lengths of guides
guide_hits <- search_index(guides_cas12, out_dir_index, distance = 2, validate = FALSE)
guide_hits_table <- read.table(guide_hits, sep = ",", header = TRUE)
guide_hits_table
```

## Short crisprVerse example

First we replicate the default bowtie run from crisprVerse vignette,
then we show how to run CHOPOFF to get similar results.

```{r crisprVerse_simple}
# Load required libraries
{
library(crisprDesign)
library(BSgenome.Hsapiens.UCSC.hg38)
library(crisprBase)
library(Rbowtie)
library(crisprCHOPOFF)
}
# Run crisprVerse example (find guide candidates in human gene: IQSEC3)
{
  data(SpCas9, package="crisprBase")
  SpCas9
  gr <- queryTxObject(txObject=grListExample,
                      featureType="cds",
                      queryColumn="gene_symbol",
                      queryValue="IQSEC3")[1] # exon 1
  guideSet <- findSpacers(gr,
                          bsgenome=BSgenome.Hsapiens.UCSC.hg38,
                          crisprNuclease=SpCas9)
  
  set.seed(10)
  guideSet <- guideSet[sample(seq_along((guideSet)), 20)] # Sample a subset
  
  guideSet <- addSequenceFeatures(guideSet) # Branching point for CHOPOFF/bowtie
}

# The offtarget detection using Bowtie:
{
  # The offtarget detection using bowtie
  fasta <- system.file(package="crisprDesign", "fasta/chr12.fa")
  outdir <- tempdir()
  Rbowtie::bowtie_build(fasta,
                        outdir=outdir,
                        force=TRUE,
                        prefix="chr12")
  bowtie_index <- file.path(outdir, "chr12")
  
  guideSetBowTie <- addSpacerAlignments(guideSet,
                                  txObject=grListExample,
                                  aligner_index=bowtie_index,
                                  bsgenome=BSgenome.Hsapiens.UCSC.hg38,
                                  n_mismatches=2)
  # Off target score
  guideSetBowTie <- addOffTargetScores(guideSetBowTie)
  # On target score
  guideSetBowTie <- addOnTargetScores(guideSetBowTie, methods="crisprater")
}
# The offtarget detection using CHOPOFF:
{
  # Input parameters
name <- "human_chr12_CHOPOFF_CAS9"
out_dir_index <- file.path(tempdir(), name)
no_fasta_index <- !file.exists(paste0(fasta, ".fai"))
if (no_fasta_index) Rsamtools::indexFa(fasta)

build_index(name, fasta, out_dir_index, distance = 2, validate = FALSE)
guideSetCHOPOFF <- addSpacerAlignmentsCHOPOFF(guideSet, out_dir_index, 
                                              distance = 2, validate = FALSE)
# On target score (This is valid)
guideSetCHOPOFF <- addOnTargetScores(guideSetCHOPOFF, methods="crisprater")
}

# Results to compare
guideSetBowTie
guideSetCHOPOFF

gsCHOPPOFF_n0 <- guideSetCHOPOFF[guideSetCHOPOFF$n1 == 0 & guideSetCHOPOFF$n2 == 0]
CHOPOFF_gr <- unlist(gsCHOPPOFF_n0$alignments, use.names = FALSE)
bowtie_gr <-unlist(guideSetBowTie$alignments, use.names = FALSE)
identical(ranges(gsCHOPPOFF_n0), ranges(guideSetBowTie[names(gsCHOPPOFF_n0)]))
# All 0 distances matches coordinates
# The others do not, since they do not support bulges.
```
## Full crisprVerse example

First we replicate the default bowtie run from crisprVerse vignette,
then we show how to run CHOPOFF to get similar results.
This example is better to run through, since it contains some additional
analysis for understanding.

```{r crisprVerse_full}
# Load required libraries
{
library(crisprDesign)
library(BSgenome.Hsapiens.UCSC.hg38)
library(crisprBase)
library(Rbowtie)
library(crisprCHOPOFF)
}
# Run crisprVerse example (find guide candidates in human gene: IQSEC3)
{
  data(SpCas9, package="crisprBase")
  SpCas9
  gr <- queryTxObject(txObject=grListExample,
                      featureType="cds",
                      queryColumn="gene_symbol",
                      queryValue="IQSEC3")[1] # exon 1
  guideSet <- findSpacers(gr, 
                          bsgenome=BSgenome.Hsapiens.UCSC.hg38,
                          crisprNuclease=SpCas9)
  guideSet
  
  set.seed(10)
  guideSet <- guideSet[sample(seq_along((guideSet)), 20)] # Sample a subset
  spacers(guideSet)
  protospacers(guideSet)
  pams(guideSet)
  head(pamSites(guideSet))
  head(cutSites(guideSet))
  
  guideSet <- addSequenceFeatures(guideSet) # Branching point for CHOPOFF
  head(guideSet)
}

# The offtarget detection using Bowtie:
{
  # The offtarget detection using bowtie
  fasta <- system.file(package="crisprDesign", "fasta/chr12.fa")
  outdir <- tempdir()
  Rbowtie::bowtie_build(fasta,
                        outdir=outdir,
                        force=TRUE,
                        prefix="chr12")
  bowtie_index <- file.path(outdir, "chr12")
  
  guideSetBowTie <- addSpacerAlignments(guideSet,
                                  txObject=grListExample,
                                  aligner_index=bowtie_index,
                                  bsgenome=BSgenome.Hsapiens.UCSC.hg38,
                                  n_mismatches=2)
  guideSetBowTie
  # Off target score
  guideSetBowTie <- addOffTargetScores(guideSetBowTie)
  guideSetBowTie
  table(guideSetBowTie$score_cfd)
  
  # On target score
  guideSetBowTie <- addOnTargetScores(guideSetBowTie, methods="crisprater")
  head(guideSetBowTie)
  table(guideSetBowTie$score_crisprater)
  
  # Final filter:
  guideSetFilter <- guideSetBowTie[guideSetBowTie$percentGC>=20]
  guideSetFilter <- guideSetFilter[guideSetFilter$percentGC<=80]
  guideSetFilter <- guideSetFilter[!guideSetFilter$polyT]
  selected_bowtie <- guideSetFilter[guideSetFilter$score_crisprater > 0.8] 
  selected_bowtie # Our final candidate
}
# The offtarget detection using CHOPOFF
{
name <- "human_chr12_CHOPOFF_CAS9"
out_dir_index <- file.path(tempdir(), name)
if (!file.exists(paste0(fasta, ".fai"))) Rsamtools::indexFa(fasta)
build_index(name, fasta, out_dir_index, distance = 2, validate = FALSE)

guideSetCHOPOFF <- addSpacerAlignmentsCHOPOFF(guideSet, out_dir_index, distance = 2)

# Off target score (This do not really make sense, because CHOPOFF supports bulges)
# guideSetCHOPOFF <- addOffTargetScores(guideSetCHOPOFF)
# guideSetCHOPOFF
# table(guideSetCHOPOFF$score_cfd)

# On target score (This is valid)
guideSetCHOPOFF <- addOnTargetScores(guideSetCHOPOFF, methods="crisprater")
head(guideSetCHOPOFF)
table(guideSetCHOPOFF$score_crisprater)

# Final filter:
guideSetCHOPOFFFilter <- guideSetCHOPOFF[guideSetCHOPOFF$percentGC>=20]
guideSetCHOPOFFFilter <- guideSetCHOPOFFFilter[guideSetCHOPOFFFilter$percentGC<=80]
guideSetCHOPOFFFilter <- guideSetCHOPOFFFilter[!guideSetCHOPOFFFilter$polyT]
selected_CHOPOFF <-  guideSetCHOPOFFFilter[guideSetCHOPOFFFilter$score_crisprater > 0.8] 
selected_CHOPOFF # Our final candidate
}

# Now lets compare
identical(ranges(selected_CHOPOFF), ranges(selected_bowtie)) # We pick same site

# All guides analysis
guideSetBowTie
guideSetCHOPOFF

identical(as.character(guideSetBowTie$protospacer), as.character(guideSetCHOPOFF$protospacer))

identical(ranges(guideSetBowTie), ranges(guideSetCHOPOFF))
identical(guideSetBowTie$n0, guideSetCHOPOFF$n0)
# CHOPOFF supports searching for bulges, so results do not match perfectly.

table(mcols(guideSetBowTie)[,c("n0")])
table(mcols(guideSetCHOPOFF)[,c("n0")])
table(mcols(guideSetBowTie)[,c("n2")])
table(mcols(guideSetCHOPOFF)[,c("n2")]) # <- 1 (1 distance hit), 2 (2 distance hits)
```

