---
title: "CHOPOFF GUIDE"
author: "Kornel Labun & Haakon Tjeldnes"
date: "`r BiocStyle::doc_date()`"
package: "`r pkg_ver('ORFik')`"
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{CHOPOFF GUIDE}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Introduction

Welcome

# Overview

# Full crisprVerse example

```{r crisprVerse}
# Load required libraries
library(crisprDesign)
library(BSgenome.Hsapiens.UCSC.hg38)
library(crisprBase)
library(Rbowtie)
library(crisprCHOPOFF)
library(data.table)

# Run default example (find guide candidates in human gene: IQSEC3)
data(SpCas9, package="crisprBase")
SpCas9

gr <- queryTxObject(txObject=grListExample,
                    featureType="cds",
                    queryColumn="gene_symbol",
                    queryValue="IQSEC3")
gr <- gr[1] # exon 1
bsgenome <- BSgenome.Hsapiens.UCSC.hg38
guideSet <- findSpacers(gr,
                        bsgenome=bsgenome,
                        crisprNuclease=SpCas9)
guideSet

set.seed(10)
guideSet <- guideSet[sample(seq_along((guideSet)),20)] # Sample a subset
spacers(guideSet)
protospacers(guideSet)
pams(guideSet)
head(pamSites(guideSet))
head(cutSites(guideSet))

guideSet <- addSequenceFeatures(guideSet)
head(guideSet)

guideSetTemplate <- guideSet # Branching point for CHOPOFF
# The offtarget detection using bowtie
fasta <- system.file(package="crisprDesign", "fasta/chr12.fa")
outdir <- tempdir()
Rbowtie::bowtie_build(fasta,
                      outdir=outdir,
                      force=TRUE,
                      prefix="chr12")
bowtie_index <- file.path(outdir, "chr12")

guideSet <- addSpacerAlignments(guideSet,
                                txObject=grListExample,
                                aligner_index=bowtie_index,
                                bsgenome=bsgenome,
                                n_mismatches=2)
guideSetBowTie <- guideSet # To compare with CHOPOFF output
guideSet
# Off target score
guideSet <- addOffTargetScores(guideSet)
guideSet
table(guideSet$score_cfd)

# On target score
guideSet <- addOnTargetScores(guideSet, methods="crisprater")
head(guideSet)
table(guideSet$score_crisprater)

# Final filter:
guideSetFilter <- guideSet[guideSet$percentGC>=20]
guideSetFilter <- guideSetFilter[guideSetFilter$percentGC<=80]
guideSetFilter <- guideSetFilter[!guideSetFilter$polyT]
selected_bowtie <- guideSetFilter[guideSetFilter$score_crisprater > 0.8] 
selected_bowtie # Our final candidate

# The offtarget detection using CHOPOFF

name <- "human_chr12_CHOPOFF_CAS9"
out_dir_index <- file.path(tempdir(), name)
if (!file.exists(paste0(fasta, ".fai"))) Rsamtools::indexFa(fasta)
build_index(name, fasta, out_dir_index, distance = 2, validate = FALSE)
guideSetCHOPOFF <- guideSetTemplate
guide_hits <- search_index(guideSetCHOPOFF, out_dir_index, distance = 2, validate = FALSE)

# Read in CHOPOFF hits from guide search
guide_hits_table <- read.table(guide_hits, sep = ",", header = TRUE)

# Convert back to crisprVerse GuideSet object
guide_hits_table <- as.data.table(guide_hits_table)
guide_hits_table[, alignments := paste0(chromosome, ":", start, ":", strand)]
# Make mismatch scoring object
alignments <- data.table::dcast(as.data.table(guide_hits_table), guide ~ distance)
colnames(alignments)[-1] <- paste0("n", colnames(alignments)[-1])
matching <- match(as.character(guideSetCHOPOFF$protospacer),  alignments$guide)
alignments_matched <- alignments[matching, ]

mcols(guideSetCHOPOFF) <- cbind(mcols(guideSetCHOPOFF), alignments_matched[, -1])
# Now make the GuideSet [["alignments]] mcol needed for scoring
mcols(guideSet)[["alignments"]] # <- This is what we want to make
matched_dups_a <- match(alignments_matched$guide, guide_hits_table$guide)
matched_dups_g <- match(guide_hits_table$guide, alignments_matched$guide)
ordering <- order(matched_dups_g)
guides_hits_table_ordered <- guide_hits_table[ordering,]
N <- guides_hits_table_ordered[, .N, by = guide]$N
dups_total <- rep.int(seq(length(guideSetCHOPOFF)), times = N)
alignments_grl <- guideSetCHOPOFF[dups_total]
# We should have defined spacer as: guides_hits_table_ordered$alignment_guide
# But crisprVerse does not support this, so we use guide.
mcols(alignments_grl) <- cbind(spacer = guides_hits_table_ordered$guide, 
                               mcols(alignments_grl)[c("protospacer", "pam", "pam_site")],
                               n_mismatches = guides_hits_table_ordered$distance,
                               canonical = TRUE, 
                               mcols(alignments_grl)["cut_site"])
alignments_grl <- split(alignments_grl, names(alignments_grl))
mcols(guideSetCHOPOFF)[["alignments"]] <- alignments_grl

# Off target score
guideSetCHOPOFF <- addOffTargetScores(guideSetCHOPOFF)
guideSetCHOPOFF
table(guideSetCHOPOFF$score_cfd)

# On target score
guideSetCHOPOFF <- addOnTargetScores(guideSetCHOPOFF, methods="crisprater")
head(guideSetCHOPOFF)
table(guideSetCHOPOFF$score_crisprater)

# Final filter:
guideSetCHOPOFFFilter <- guideSetCHOPOFF[guideSetCHOPOFF$percentGC>=20]
guideSetCHOPOFFFilter <- guideSetCHOPOFFFilter[guideSetCHOPOFFFilter$percentGC<=80]
guideSetCHOPOFFFilter <- guideSetCHOPOFFFilter[!guideSetCHOPOFFFilter$polyT]
selected_CHOPOFF <-  guideSetCHOPOFFFilter[guideSetCHOPOFFFilter$score_crisprater > 0.8] 
selected_CHOPOFF # Our final candidate

# Now lets compare
identical(ranges(selected_CHOPOFF), ranges(selected_bowtie)) # We pick same site

# All guides analysis
guideSetBowTie
guideSetCHOPOFF

identical(as.character(guideSetBowTie$protospacer), as.character(guideSetCHOPOFF$protospacer))

identical(ranges(guideSetBowTie), ranges(guideSetCHOPOFF))
identical(guideSetBowTie$n0, guideSetCHOPOFF$n0)
# CHOPOFF supports searching for bulges, so results do not match.

table(mcols(guideSetBowTie)[,c("n0")])
table(mcols(guideSetCHOPOFF)[,c("n0")])
table(mcols(guideSetBowTie)[,c("n2")])
table(mcols(guideSetCHOPOFF)[,c("n2")]) # <- 1 (1 distance hit), 2 (2 distance hits)


```

